src/clib/pio_msg.c

```diff
@@ -378,11 +399,20 @@ int inq_att_handler(iosystem_desc_t *ios)
         return pio_err(ios, NULL, PIO_ENOMEM, __FILE__, __LINE__);
     if ((mpierr = MPI_Bcast((void *)name, namelen + 1, MPI_CHAR, ios->compmaster,
                             ios->intercomm)))
+    {
+        free(name);
         return check_mpi2(ios, NULL, mpierr, __FILE__, __LINE__);
+    }
     if ((mpierr = MPI_Bcast(&xtype_present, 1, MPI_CHAR, 0, ios->intercomm)))
+    {
+        free(name);
         return check_mpi2(ios, NULL, mpierr, __FILE__, __LINE__);
+    }
     if ((mpierr = MPI_Bcast(&len_present, 1, MPI_CHAR, 0, ios->intercomm)))
+    {
+        free(name);
         return check_mpi2(ios, NULL, mpierr, __FILE__, __LINE__);
+    }
```

Comment:
The issue is that in current code, we have a lot of places where we would return earlier on an MPI error
```
if ((mpierr = MPI_Bcast(...))
    return check_mpi(NULL, mpierr, __FILE__, __LINE__);
```
If you do not like the above code, we can change to the new structure you suggested.